platform/ios/src/MGLMapView.mm

```diff
@@ -3024,50 +3090,65 @@ - (void)locationManager:(__unused CLLocationManager *)manager didUpdateLocations
+                    __weak MGLMapView *weakSelf = self;
+                    [self _flyToCamera:camera withDuration:animated ? -1 : 0 peakAltitude:-1 completionHandler:^{
+                        MGLMapView *strongSelf = weakSelf;
+                        strongSelf.userTrackingState = MGLUserTrackingStateChanged;
+                    }];
+                }
             }
-            else
+            else if (self.userTrackingState == MGLUserTrackingStatePossible)
             {
                 // otherwise re-center and zoom in to near accuracy confidence
                 //
-                float delta = (newLocation.horizontalAccuracy / 110000) * 1.2; // approx. meter per degree latitude, plus some margin
-
-                CLLocationCoordinate2D desiredSouthWest = CLLocationCoordinate2DMake(newLocation.coordinate.latitude  - delta,
-                                                                                     newLocation.coordinate.longitude - delta);
-
-                CLLocationCoordinate2D desiredNorthEast = CLLocationCoordinate2DMake(newLocation.coordinate.latitude  + delta,
-                                                                                     newLocation.coordinate.longitude + delta);
-
-                CGFloat pixelRadius = fminf(self.bounds.size.width, self.bounds.size.height) / 2;
-
-                CLLocationCoordinate2D actualSouthWest = [self convertPoint:CGPointMake(userLocationPoint.x - pixelRadius,
-                                                                                        userLocationPoint.y - pixelRadius)
-                                                       toCoordinateFromView:self];
-
-                CLLocationCoordinate2D actualNorthEast = [self convertPoint:CGPointMake(userLocationPoint.x + pixelRadius,
-                                                                                        userLocationPoint.y + pixelRadius)
-                                                       toCoordinateFromView:self];
-
-                if (desiredNorthEast.latitude  != actualNorthEast.latitude  ||
-                    desiredNorthEast.longitude != actualNorthEast.longitude ||
-                    desiredSouthWest.latitude  != actualSouthWest.latitude  ||
-                    desiredSouthWest.longitude != actualSouthWest.longitude)
-                {
-                    // assumes we won't disrupt tracking mode
-                    [self setVisibleCoordinateBounds:MGLCoordinateBoundsMake(desiredSouthWest, desiredNorthEast) edgePadding:UIEdgeInsetsZero direction:course animated:YES];
-                    [self unrotateIfNeededAnimated:YES];
-                }
+                self.userTrackingState = MGLUserTrackingStateBegan;
+                
+                MGLMapCamera *camera = self.camera;
+                camera.centerCoordinate = self.userLocation.location.coordinate;
+                camera.heading = course;
+                camera.altitude = newLocation.horizontalAccuracy;
```

Comment:
You're right, it does always seem to zoom into z18 for some reason. Perhaps it's because iOS-specific code assumes a field of view of 30 degrees, which is less than what some mbgl calculations assume.